lib/fast_jsonapi/serialization_core.rb

```diff
@@ -65,55 +64,102 @@ def meta_hash(record, params = {})
+            record_hash_pairs = uncached_by_opts.values.reduce({}, :merge).merge!(cache_hits)
+
+            # push all record_hash to batch loader context
+            batch_params.each do |b_param|
+              loader.call(b_param, record_hash_pairs[b_param[:cache_key]])
+            end
           end
-          record_hash[:relationships] = (record_hash[:relationships] || {}).merge(relationships_hash(record, uncachable_relationships_to_serialize, fieldset, includes_list, params)) if uncachable_relationships_to_serialize.present?
         else
-          record_hash = id_hash(id_from_record(record, params), record_type, true)
-          record_hash[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
-          record_hash[:relationships] = relationships_hash(record, nil, fieldset, includes_list, params) if relationships_to_serialize.present?
-          record_hash[:links] = links_hash(record, params) if data_links.present?
+          record_h = id_hash(id_from_record(record, params), record_type, true)
+          record_h[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
+          record_h[:relationships] = relationships_hash(record, nil, fieldset, includes_list, params) if relationships_to_serialize.present?
+          record_h[:links] = links_hash(record, params) if data_links.present?
+          record_h[:meta] = meta_hash(record, params) if meta_to_serialize.present?
+          record_h
         end
-
-        record_hash[:meta] = meta_hash(record, params) if meta_to_serialize.present?
-        record_hash
       end
 
       # Cache options helper. Use it to adapt cache keys/rules.
       #
       # If a fieldset is specified, it modifies the namespace to include the
       # fields from the fieldset.
       #
+      # Allow namespace passed in a proc to generate unique key base on dynamic params
+      #
       # @param options [Hash] default cache options
       # @param fieldset [Array, nil] passed fieldset values
       # @param includes_list [Array, nil] passed included values
       # @param params [Hash] the serializer params
       #
       # @return [Hash] processed options hash
       # rubocop:disable Lint/UnusedMethodArgument
-      def record_cache_options(options, fieldset, includes_list, params)
-        return options unless fieldset
-
+      def record_cache_options(record, options, fieldset, includes_list, params)
         options = options ? options.dup : {}
+
+        if options[:namespace].is_a?(Proc)
+          options[:namespace] = FastJsonapi.call_proc(options[:namespace], record, params)
+        end
+
         options[:namespace] ||= 'jsonapi-serializer'
 
-        fieldset_key = fieldset.join('_')
+        return options unless fieldset
 
         # Use a fixed-length fieldset key if the current length is more than
         # the length of a SHA1 digest
+        fieldset_key = fieldset.join('-')
```

Comment:
why?

Reply:
Why use dash over underscore? It's a matter of taste, also want to consistent with cache_key style like `meal_instances/321724-20210617211745378854`.

Reply:
not worth a breaking change for the gem